lib/resolver.js

```diff
@@ -1,237 +1,194 @@
+  });
+};
 
 /**
- * The heavy lifter. This is a recursive method that will traverse over each key in
- * the given `schema` and resolve the values until there are no $ref occurrences left.
+ * The heavy lifter. This is a recursive function that will traverse over each key in
+ * the given schemas (schemas) and builds a new schema, where all $refs are
+ * resolved.
  *
- * @param {Object} schema - a valid JSON schema object
- * @param {String} context - schema ID of the current resolving context
- * @param {String} [prop] - Used in a recursive context, specifically when a `$ref` is found and the result of the `$ref` needs to be assigned to the original property
- * @param {Object} [stack] - The parent object that contains the above `prop` when in the `$ref` context, so we can assign the result to the parent.
- * @returns {*}
+ * @param {String} schemaId - top level schema ID, context for internal resolutions
+ * @param {String} path - JSON pointer of the schemas node that we are dereferencing
+ * @param {Object} schemas - schemas, our "recipe" for building new schemas
+ * @param {Number} debugLevel
+ * @returns {Object} - resolved schema, without $refs
  */
-Resolver.prototype.dereferenceSchema = function (schema, context, prop, stack) {
-  // If the value is not an object, we've reach the end of the line, so just return the value
-  if (!_.isPlainObject(schema)) {
-    return schema;
-  }
-
-  // Loop through the object and recursively resolve each value
-  _.each(schema, function(item, property){
-    var resolved;
-    // Found a sub-schema
-    if (property === '$ref') {
-      this._debug(3, 'Found $ref: %s in %s - Resolving...', item, context);
-      // Resolve schema or definition reference (throws if it can't find it)
-      resolved = this._resolvePointer(item, context);
-      // Assign the resolved reference as the schema for this prop/stack loop.
-      // Pass along the resolved ID if it's a valid schema, to
-      // force a context change when recursing
-      schema = this.dereferenceSchema(resolved, this._normalizeReference(item, context), prop, stack);
-      // Quit this loop once we've found a `$ref`
-      return false;
+var dereferenceSchema = function(schemaId, path, schemas, counters, debugLevel) {
+  return _.mapValues(resolveReference(schemaId, path, schemas), function(value, name) {
+    if (_.isString(value)) {
+      // resolving $refs
+      if (name === '$ref') {
+
+        // ############### CIRCULAR REFERENCE DETECTION ################## //
+        var normRef = normalizeReference(schemaId, path, schemas);
+        var refNode = pointer.get(schemas[normRef.schemaId], normRef.path);
+        var nodePath = normRef.schemaId + normRef.path;
+
+        if (!_.has(counters, nodePath)) {
+          counters[nodePath] = 0;
+        }
+        // break resolving if $ref in this subschema has been already resolved
+        if (counters[nodePath] > RECURSION_LEVELS) {
+          if (debugLevel > 0) {
```

Comment:
This helper seems to be working only in a class context, so I can't. But I will update and unify it in an another PR.